assets/js/components/PermissionsModal/AuthenticatedPermissionsModal.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * AuthenticatedPermissionsModal component.
+ *
+ * Site Kit by Google, Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useEffect, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
+import { CORE_LOCATION } from '../../googlesitekit/datastore/location/constants';
+import { snapshotAllStores } from '../../googlesitekit/data/create-snapshot-store';
+import Dialog from '../Dialog';
+import Portal from '../Portal';
+const { useSelect, useDispatch, useRegistry } = Data;
+
+const AuthenticatedPermissionsModal = () => {
+	const registry = useRegistry();
+	const permissionsError = useSelect( ( select ) =>
+		select( CORE_USER ).getPermissionScopeError()
+	);
+	const connectURL = useSelect( ( select ) =>
+		select( CORE_USER ).getConnectURL( {
+			additionalScopes: permissionsError?.data?.scopes,
+			redirectURL: global.location.href,
+		} )
+	);
+
+	const { clearPermissionScopeError } = useDispatch( CORE_USER );
+	const { navigateTo } = useDispatch( CORE_LOCATION );
+
+	const onCancel = useCallback( () => {
+		clearPermissionScopeError();
+	}, [ clearPermissionScopeError ] );
+
+	const onConfirm = useCallback( async () => {
+		// If we have a datastores to snapshot before navigating away to the
+		// authorization page, do that first.
+		await snapshotAllStores( registry );
+		navigateTo( connectURL );
+	}, [ registry, connectURL, navigateTo ] );
+
+	useEffect( () => {
+		// If error has flag to skip the modal, redirect to the authorization
+		// page immediately without prompting the user, essentially short-
+		// circuiting to the confirm step.
+		const confirmIfSkipModal = async () => {
+			if (
+				permissionsError?.data?.skipModal &&
+				permissionsError?.data?.scopes?.length
+			) {
+				await onConfirm();
+			}
+		};
+		confirmIfSkipModal();
+	}, [ onConfirm, permissionsError ] );
+
+	if ( ! permissionsError ) {
```

Comment:
It looks like this can potentially return `undefined`, which I was assuming is from it resolving and returning `undefined` during that, but it appears that actually isn't the case… it's either `null` or an error.

But the docs say it can return `undefined`:

https://github.com/google/site-kit-wp/blob/d3aed2ee97a0b122b78fcc304041c0bf85dc6d92/assets/js/googlesitekit/datastore/user/permissions.js#L147-L159

This isn't related to your change, but we should update the JSDoc for `getPermissionScopeError` so it doesn't trip anyone else up. Looks like it should be `@return {(Object|undefined)}` not `@return {(Object|null)}`

Reply:
Worse: I think it was me who wrong that JSDoc. Shame! 😆